src/distance.cpp

```diff
@@ -0,0 +1,620 @@
+    //Get the length of a chain including length of last node
+    return prefix_sum->back();
+}
+
+
+int64_t calculateIndex(DistanceIndex* di, VG* graph, SnarlManager* sm, 
+         const Chain* chain) {
+    /*Add chain to distance index and recursively add all distances in 
+      snarls contained within chain
+    */
+    auto cmp = [] (pair<const handle_t*,int64_t> x,
+                                           pair<const handle_t*,int64_t> y) {
+        //Comparison function for the priority of a pair of handle, distance
+        return (x.second < y.second); 
+    };
+
+    vector<int64_t> chain_prefix_sum (1,0); //initialize to [0]
+
+    unordered_map<id_t, size_t> chain_snarl_to_index;
+    for (int i = 0; i < chain->size(); i++) { 
+        //for each snarl in the chain TODO ChainIterator?
+        const Snarl* snarl = chain->at(i);
+        chain_snarl_to_index.insert( make_pair<id_t, size_t>
+                     (snarl->start().node_id(), chain_snarl_to_index.size()));
+        //TODO use visit as key to identify a node+direction? or node id? 
+        //Get all nodes in snarl includes both boundary nodes of child snarls
+        pair<unordered_set<Node*>, unordered_set<Edge*>> contents = 
+                        sm->shallow_contents(snarl, *graph, true);
+        unordered_set<Node*> allNodes = contents.first;
+        /*TODO better way of getting all the nodes in the snarl? Currently
+          don't include end boundary node of a child snarl in distance 
+          matrix because I don't think I need them?
+        */ 
+
+         // Create SnarlDistance object with all nodes in this snarl
+        unordered_map <pair<id_t, bool>, size_t> vti;
+        size_t snarlIndex = 0;  
+        for (unordered_set<Node*>::iterator n = allNodes.begin(); 
+            n != allNodes.end(); ++n) {
+
+            const Node* node = *n;
+            id_t nID = node->id();
+
+            vti.insert(make_pair(make_pair(nID, true), vti.size()));
+            vti.insert(make_pair(make_pair(nID, false), vti.size()));
+        }
+        int size = vti.size();
+        //Initialize all distances to -1
+        vector<int64_t> distance_obj (size*size, -1);
+        SnarlDistances sd = SnarlDistances(&vti, &distance_obj);
+        //add this snarl distance to distance index
+        di->sd.insert(make_pair<id_t, SnarlDistances*>
+                                                (snarl->start().node_id(),&sd));
+
+        //Create a NetGraph for current snarl
+        NetGraph ng = 
+            NetGraph(snarl->start(), snarl->end(), sm->chains_of(snarl), graph);
+
+
+        // For each node in snarl calculate distance to every reachable node
+        for (unordered_set<Node*>::iterator n = allNodes.begin(); 
+                                                     n != allNodes.end(); ++n){
+        bool bools [2] = {true, false};
+        for (bool rev : bools) {
+        //TODO Better way to loop over all nodes/direction??
```

Comment:
Formally speaking, no. This makes sense as a method of iteration. Stylistically speaking, I might prefer
```
for (Node* n : allNodes) {
    for (bool rev : {true, false}) {
        ...
    }
}
```